fix(router): experimental custom error handler

dai-shi opened the pull request:
follow up #1247

experimental router api for #1239

rmarscher commented on packages/waku/src/router/client.ts:
```diff
@@ -342,6 +346,58 @@ export class ErrorBoundary extends Component<
+  );
+};
+
+class CustomErrorHandler extends Component<
```
Can we add a way to pass in components to use instead of the defaults? I want to be able to define my own NotFound component, for example. Maybe we can pass it through as a `<Router />` prop or another method?

https://github.com/vercel/next.js/blob/v15.1.0/packages/next/src/client/components/http-access-fallback/error-boundary.tsx#L169-L177

dai-shi replied:
I think you would like to use an error that isn't captured by CustomErrorHandler. Let's revisit it as our design doesn't seem very solid.

rmarscher replied:
This comment was about being able to create different element output than `<html><body><h1>Not Found</h1></body></html>` component when there is a stream. If it can render a [404 page component if available](https://github.com/dai-shi/waku/pull/1261/files#r1967194656), that works.

rmarscher reviewed (commented):
Thank you for working on this.

Do we need to also make updates here on the server-side? https://github.com/dai-shi/waku/blob/main/packages/waku/src/lib/middleware/handler.ts#L162-L192

I think it would ideally render a new RSC stream. We have Hono fallback 404 html that can be configured somewhere. Without a new RSC render, I think new NotFound component will only happen when navigating on the client.

To test, I added `unstable_notFound()` to the pages/about.tsx in the 01_template. I updated `pages/index.tsx` and `pages/about.tsx` be dynamic while leaving `_layout.tsx` as static.

I may have done something wrong, but I'm not seeing the Not Found custom error getting caught by the custom error boundary even with client-side routing.

dai-shi commented on packages/waku/src/router/client.ts:
```diff
@@ -342,6 +346,58 @@ export class ErrorBoundary extends Component<
   }
 }
 
+const NotFound = () => {
+  // FIXME totally unsure if this is a desired behavior
+  useEffect(() => {
+    window.location.replace('/404');
+  });
+  return createElement(
+    'html',
+    null,
+    createElement('body', null, createElement('h1', null, 'Not Found')),
+  );
+};
```
Without the stream, we show `src/pages/404.txt` if available, so I would like to keep it consistent. However, there's also `public/public.html`. So, I guess nothing works ideally as of now. We probably need to define the expected behavior from scratch.

It feels impossible to behaves the same between with stream and without it.

dai-shi commented on packages/waku/src/router/client.ts:
```diff
@@ -342,6 +346,58 @@ export class ErrorBoundary extends Component<
+  useEffect(() => {
+    window.location.replace(loc);
+  });
+  return createElement(
+    'html',
+    null,
+    createElement('body', null, createElement('a', { href: loc }, 'Redirect')),
+  );
```
Hm, I think we need to put `CustomErrorHandler` in `Router` then.

dai-shi commented:
> Do we need to also make updates here on the server-side? https://github.com/dai-shi/waku/blob/main/packages/waku/src/lib/middleware/handler.ts#L162-L192

I don't believe so, but we might be having a different assumption. And, I don't believe my assumption is correct.

> I think it would ideally render a new RSC stream. We have Hono fallback 404 html that can be configured somewhere. Without a new RSC render, I think new NotFound component will only happen when navigating on the client.

I think it should show `public/404.html`, but not sure.

> To test, I added `unstable_notFound()` to the pages/about.tsx in the 01_template. I updated `pages/index.tsx` and `pages/about.tsx` be dynamic while leaving `_layout.tsx` as static.
> 
> I may have done something wrong, but I'm not seeing the Not Found custom error getting caught by the custom error boundary even with client-side routing.

The custom error boundary would work only when you throw an error in server-side suspense. Did you add Suspense on the server?

Oh, this seems pretty complicated. Thanks for giving me feedback. I would like to make Waku as thin as possible, but also with a good default. Hope we find a balanced solution.

dai-shi commented:
> Do we need to also make updates here on the server-side? https://github.com/dai-shi/waku/blob/main/packages/waku/src/lib/middleware/handler.ts#L162-L192

Actually, I think I misunderstood something. Will fix it.



> Maybe `error.digest` is the only way to pass custom error information?

Oh... That's my oversight. I'll look into it.

dai-shi commented:
```js
> const e = new Error('err')
undefined
> e.foo=1
1
> const e2 = structuredClone(e)
undefined
> e2.foo
undefined
```
I see.

dai-shi commented:
serializeErrorValue
https://github.com/facebook/react/blob/25677265038b89c1ee3000e0669339ed160d9d75/packages/react-server/src/ReactFlightServer.js#L3096-L3126

It doesn't have any `.digest` handling. It looks like some React feature.
